### ISSUE TITLE
Support Arbitrary Protocols

### ISSUE THREAD

**Issue by HalosGhost** [mvdan/xurls#1]
It looks like ``xurls`` only checks for ``https?`` which grabs ``http`` and ``https`` instead of allowing arbitrary protocols (e.g., ``file://``, ``ftp://``, ``steam://``, etc.).

Any interest in adding support for arbitrary protocols?

**Comment by mvdan** [comment]
Sure! I just didn't want to overcomplicate things for the first implementation.

The thing is that the whole regex currently revolves around web links. If we were to add more protocols, the regex would be split into two:

* text without `foo://` but which ressembles a web link, e.g. by having `.com` in it (implemented)
* text with `foo://` and a following array of allowed characters (implemented for `https?` only)

I guess my question then is, are the arrays of allowed characters uniform across protocols? I'm thinking not, in which case we could have a per-protocol regex or just make a superset of all the regexes and oversimplify that way.

Thoughts?

**Comment by mvdan** [comment]
We could also have a mix: if we have a regex for the found protocol, use that. If not, fall back to the generic regex that would exclude whitespaces and other illegal characters.

**Comment by HalosGhost** [comment]
Personally, I would be under the impression that all of them have the same limitations, and I do not think it would be wrong to have the first try at implementing them be generic in that sense. If after that, it turns out that there are more stringent limitations on some protocols than others, perhaps a small dictionary of regexs would make sense with the generic implementation serving as a fallback.

**Comment by mvdan** [comment]
Cool, thanks for the feedback! I'll get it done in the following days.

I'm curious, what do you use xurls for?

**Comment by HalosGhost** [comment]
I use ``st`` for my terminal emulator paired with ``tmux`` for tabs and sessions. I use your ``xurls`` as a replacement for the old ``xurls`` perl script to grab URLs present in a tmux pane and pass them to ``dmenu``. Depending on what I use to call this process, the URL I select through ``dmenu`` is either placed in my primary selection or opened with ``$BROWSER``.

Functionally, I use your ``xurls`` for URL integration in my TE.

**Comment by mvdan** [comment]
Funny, because I also use st and tmux. And I actually wrote xurls as a replacement for the old perl version :) Funnily enough though, I haven't yet started using it in st yet. Do you use the `externalpipe` patch?

So for example `-s` for strict, as in don't try to match links that are missing the protocol part? That indeed sounds interesting. Please open a separate issue :)

Something just occured to me, what about protocols that have `:` instead of `://` like `mailto:` or `xmpp:`?

**Comment by HalosGhost** [comment]
Actually, I do not use the externalpipe patch. Rather, I use a small script (which I call ``turl``):

```sh
#!/bin/sh

get_urls () {
    tmux capture-pane -J
    tmux show-buffer | xurls | tac | dmenu -b -l 10
}

follow () {
    get_urls | xargs -r "${BROWSER}" > /dev/null
}

grab () {
    get_urls | xargs -r | xsel -i
}

if [[ "$1" == "-g" ]]; then
    grab
else
    follow
fi
```

It's really quick and not terribly robust, but it gets the job done quite well. I have it bound to two different key combinations, one with the ``-g`` flag, the other without.

As for the other protocols you mention, I do not see a problem with their being included. I feel like the simplest way of doing this is having a small dictionary of known protocols which operate without a ``://`` and then using arbitrary text matching for anything that does have a ``://``. Thoughts?

**Comment by mvdan** [comment]
Do you keep that code anywhere? I might start using it.

Yep, manually matching the ones that don't have the extra `//` is probably best since only matching `foo:` would cause way too many false positives.

**Comment by HalosGhost** [comment]
I keep ``turl`` and all my other simple personal utils in my [.bin repo](https://github.com/HalosGhost/.bin).

You can also see my ``tmux`` configuration in my [.dotfiles repo](https://github.com/HalosGhost/.dotfiles).

**Comment by mvdan** [comment]
I haven't forgotten about your suggestions, I'm just doing cleanup and refactoring before adding large new features :)

**Comment by HalosGhost** [comment]
Completely understandable. Though, now is a good time to also suggest that `sms:` be added to the special no-slashes list of protocols to match. :)

**Comment by mvdan** [comment]
I've come up with a conflict with the current usage of the command with the `-e` flag for matching emails instead of http urls.

After the rewriting, how will it behave? It made sense before as I only had regex for http urls and email addresses, but now if I add support for, say, `sms:` like you suggest, it falls under neither category.

Long story short, one option is to:
* Separate `mailto:` and the current email address regex into `-e`
* The rest, including `sms:`, `xmpp:` et al fall under the default behavior, which is to match "urls" (it never said they had to be http urls)

Another option is to:
* Get rid of `-e` and always print all found urls, leaving it to the user or another program like sed or grep to filter out the results

Thoughts?

**Comment by HalosGhost** [comment]
Personally, I'd recommend removing the `-e` flag entirely. Since you've opted to do arbitrary string matching for the general case, the most obvious flag to replace it with (a flag that takes a comma-separated list of protocols as a filter) isn't as simple. But, in the future, something like that might be a great idea.

In the meantime though, I think removing `-e` makes a lot more sense than trying to keep it around and making other things more complicated.

**Comment by mvdan** [comment]
Okay, master now works with any scheme followed by `://`. Now adding support for a limited set of schemes that use `:` as discussed.

**Comment by HalosGhost** [comment]
via eced9a4

**Comment by mvdan** [comment]
Well, I was going to close it once I added `mailto:`, `sms:` and the rest :)

**Comment by HalosGhost** [comment]
:+1:

**Comment by mvdan** [comment]
With that fixed, do you have any more suggestions? :)

**Comment by HalosGhost** [comment]
I'll need to think about it; this is already improved so much. I guess the next possibility would be for users to be able to specify a comma-separated list of protocols to match; e.g., ``xurls -m 'mailto:,sms:,https://'``

passing text through the above command would only match URLs which are in the list. This would require some major refactoring though, I think…

**Comment by mvdan** [comment]
Wouldn't this be possible via a simple pipe through `grep "^(mailto:|sms:|https://).*"?

**Comment by mvdan** [comment]
Granted that the strict mode with `-s` could also be done via grep, but at least that's more generic and easy to implement/understand.

**Comment by HalosGhost** [comment]
Sure, it's accomplishable through external filters; but relying on external tools (and a pipe) to implement functionality as simple as narrowing the matches seems like an unfortunate decision.

But, adding this could be quite tricky without a lot of refactoring.

**Comment by mvdan** [comment]
It actually wouldn't.

Just one thing pops up in my head. I can find all urls, all urls with a scheme, and all urls with a scheme from a limited set of schemes (if this option was implemented). How would I tell it to "match all email addresses, either with or without `mailto:` in them"?

**Comment by mvdan** [comment]
In other words, I want to think this through a bit before implementing it right away. Please open a new issue per feature and this way we can keep discussion separate.

**Comment by HalosGhost** [comment]
of course; I replied here only because you asked me what other things I thought could be helpful here. I'll mull it over in more depth and do a more complete write-up when I open a new issue.

**Comment by mvdan** [comment]
Ah sure! I didn't mean it that way, just reminding myself not to ramble on random stuff on a closed issue.

**Comment by mvdan** [comment]
@HalosGhost I have an idea as to how to implement this. There's just one thing I'm not sure about - would -m "foo,bar" imply -s, error without -s or work even without -s?

To me, limiting to e.g. "mailto:" but still accepting www.foo.bar seems wrong, so I'd probably go with -m implying -s.

**Comment by HalosGhost** [comment]
I agree that having ``-m`` defaulting to strict matching makes perfect sense. But, may I offer an alternative implementation? Perhaps instead of having ``-m`` imply ``-s``, it would make more sense for ``-s`` to become default behavior and then offer something like a ``--relaxed`` to make it more lenient. Then, if for some reason people really want to match against really odd things, they could do (e.g.,) ``xurls -rm "foo"``

This is just throwing it out there though. If you do not like that idea, having ``-m`` imply ``-s`` is a fine fallback.

**Comment by mvdan** [comment]
I'm thinking that I would use -a for aggressive instead of -r for relaxed. Agree with you, the now default behaviour is silly in the real world where most urls have proper schemes and false positives would annoy people.

So -a (or -r) for said aggressive behaviour and -m for matching schemes.

**Comment by mvdan** [comment]
Let me take the -a thing back. Strict and Relaxed modes make more sense than Normal and Aggressive.

**Comment by mvdan** [comment]
Also, for -m, I have three ideas:

* -m <pattern>, where <pattern> can be in the form `(https?://|ftps?://)` and is used verbatim in the regex
* -m <cslist>, where <cslist> can be in the form `https?://,ftps?://' and is split and re-joined in a regex OR
* like the second option, but forbidding regex (i.e. use `http://,https://,ftp://,ftps://` instead). Don't see how or why I should forbid regex, though.

I'm inclined to go with the first option since it's the simplest and most versatile of all options, although it might overcomplicate things for really simple use cases.

We could drop the outermost parenthesis though, and have `https?://|ftps?://` instead of `(https?://|ftps?://)`. I think I'll go with that.

**Comment by mvdan** [comment]
(Note that `(https?://|ftps?://)` would still work, of course.)

**Comment by HalosGhost** [comment]
I like the first option with the ``()`` being optional as you stipulated. Also, is the plan now to make default behavior strict and have a switch to enable relaxation mode?

**Comment by mvdan** [comment]
Yep, it's already the default behaviour in master. I'm now implementing -m.

**Comment by mvdan** [comment]
``` 
% echo "http://foo.bar foo:bar ftps://adfja" | ./xurls -m "https?://"
http://foo.bar
```
:)

**Comment by mvdan** [comment]
All done. Test away. I'll do a release in a few days if I'm happy with how things are.

On the plus side, nearly no refactoring was needed. And since no splitting/appending happens, there's no heavy usage of the heap either.

**Comment by HalosGhost** [comment]
\o/
